lib/cid.ex

```diff
@@ -1,38 +1,87 @@
-  getting "confused" this might not be required, but is to match the original
-  "Hits" implementation.
+  Returns a CID that identical to one returned by IPFS if given the same data.
+  Can take a String, Map or Struct as an argument.
 
-  ## Parameters
+  ## Examples
 
-  - input: String the string to be hashed.
-  - length: Number the length of string required
+    iex> Cid.cid("hello")
+    "zb2rhcc1wJn2GHDLT2YkmPq5b69cXc2xfRZZmyufbjFUfBkxr"
 
-  Returns String hash of desired length.
+    iex> Cid.cid(%{key: "value"})
+    "zb2rhkN6szWhAmBFjjP8RSczv2YVNLnG1tz1Q7FyfEp8LssNZ"
   """
-  def make(input) when is_map(input) do
-    input |> stringify_map_values |> make
+  def cid(value) do
+    value
+    |> create_multihash()
+    |> create_cid()
+  end
+
+  # if create_multihash is called with a struct, the struct is converted into a
+  # map and then create_multihash is called again
+  defp create_multihash(%_{} = struct) do
+    struct
+    |> Map.from_struct()
+    |> create_multihash()
   end
 
-  def make(input, length \\ 32) do
-    hash1 = :crypto.hash(:sha512, input)
-    {:ok, <<_multihash_code, _length, hash2::binary>>} = Multihash.encode(:sha2_512, hash1)
+  # if create_multihash is called with a map the map is converted into a JSON
+  # string and then create_multihash is called again
+  defp create_multihash(map) when is_map(map) do
+    map
+    |> Jason.encode!()
+    |> create_multihash()
+  end
+
+  # if create_multihash is called with a string, the string has a new line added
+  # to the end (as that's what IPFS appears to be doing based on tests), then
+  # the string is converted into a multihash
+  defp create_multihash(str) when is_binary(str) do
+    str = add_new_line(str)
+    digest = :crypto.hash(:sha256, str)
+    {:ok, multihash} = Multihash.encode(:sha2_256, digest)
 
-    hash2
-    |> Base.encode64()
-    |> String.replace(~r/[Il0oO=\/\+]/, "", global: true)
-    |> String.slice(0..(length - 1))
+    multihash
   end
 
-  def stringify_map_values(input_map) do
-    Enum.sort(Map.keys(input_map)) # sort map keys for consistent ordering
-    |> Enum.map(fn (x) -> Map.get(input_map, x) end)
-    |> Enum.join("")
+  # if create_multihash is called something that is not a string, map or struct
+  # then it returns an error.
+  defp create_multihash(_), do: {:error, "invalid data type"}
+
+  # if an error is passed in return error message
+  defp create_cid({:error, msg}), do: msg
+
+  # takes a multihash and retuns a CID
+  defp create_cid(multihash) when is_binary(multihash) do
+    multihash
+    |> create_cid_suffix()
+    |> B58.encode58()
+    |> add_multibase_prefix()
+  end
+
+  # takes a multihash and returns the suffix
+  # currently version is hardcoded to 1
+  # and multicodec-packed-content-type is hardcoded to "raw" ("U" == <<85>>)
+  # <version><multicodec-packed-content-type><multihash>
+  defp create_cid_suffix(multihash), do: <<1>> <> "U" <> multihash
```

Comment:
A small comment on the binary syntax might be useful to people unfamiliar with it